cache/contenthash/filehash_windows.go

```diff
@@ -11,9 +11,10 @@ import (
 // chmodWindowsTarEntry is used to adjust the file permissions used in tar
 // header based on the platform the archival is done.
 func chmodWindowsTarEntry(perm os.FileMode) os.FileMode {
-	perm &= 0755
-	// Add the x bit: make everything +x from windows
-	perm |= 0111
+	if perm.IsDir() {
+		// make directories +x on windows
+		perm |= 0111
+	}
 
 	return perm
 }
```

Comment:
I'm suspicious about this whole function. If I just removed it, the tests passed, but the ones that touch the filesystem are all skipped anyway, so it might be a false pass. I tried applying this test to the change-creation code in `parseChange`, but that seemed to go wonky because it was stripping the type bits, and the whole thing just felt hacky.

Reply:
https://github.com/moby/moby/commit/1a22418f9f1573ca8521831d52c8e0562cb3ef8f#diff-b71b4407aaeb57a417a5456b663c1415
https://github.com/moby/moby/commit/41eb61d5c2a44c745b11ada948034fd3bcc42db5#diff-b71b4407aaeb57a417a5456b663c1415

Not sure I understand what the previous thing is breaking.

@simonferquel

Reply:
Yeah, https://github.com/moby/moby/commit/41eb61d5c2a44c745b11ada948034fd3bcc42db5#diff-b71b4407aaeb57a417a5456b663c1415 mentions the exact problem:
```
// this 0-ed out tar flags (like link, regular file, directory marker etc.)
```

I could just adopt that solution from that change, but then the unit tests need to have two expected digests, one for 0644 files (as generated by the unit test) on Linux, and one for mysteriously-0755 files (from the same generation code) on Windows.

It's not clear why we'd want to enforce the +x on files going into the tar header (or mess with the modes at all), given that Windows seems to always read the filesystem mode as 000, irrespective of actual permissions (although I didn't actually validate that, just observed it in the debugger, and tarballs created by the in-box Windows tar). So it's not like we're going to get false cache-hits or cache-misses from not having this code at all.

Reply:
IIRC, this is to deal with build contexts coming from a Windows client, while the daemon runs on Linux (probably applies to lcow as well): the Windows FileSystem does not really map permissions well, so to make sure shell scripts etc. coming from the build context, all files are marked executable in the tarball. In old versions of the builder (prior to the context streaming and multi-stage build), there was a warning displayed by Docker CLI saying something like "Warning, building from a Windows Client makes all file in the context have 0777 permission". @thaJeztah may have older context.

Reply:
Yes, the executable bit was set in those cases, because there was no way to set that bit on a windows filesystem (and otherwise users would always have to add a `chmod +x` in the Dockerfile to make build work from both a Windows and Linux/Mac CLI

Reply:
@TBBle I think you are right and we just shouldn't do any mode manipulation at all in here. The use-case in moby and containerd looks different. If we want to ensure windows context files have the executable bit (and if they don't already) then this should be added in the context transfer, eg. similar to https://github.com/moby/buildkit/blob/3ac3a613211f78d1c6cea9a195f640220aa21387/client/solve.go#L326

Reply:
Sounds good. As I understand, this function is not currently used to actually affect anything, it's just for checksumming the content in the cache. So I'll just rip the code out.

I definitely recall that 'all files in the context have 0777 permission' error from playing with Docker and LCOW early on.